File: Python/Easy/trees/balanced_binary_tree.py
"""

Given a binary tree, determine if it is 
height-balanced.

height-balanced: A height-balanced binary tree is a 
binary tree in which the depth of the two subtrees 
of every node never differs by more than one.

"""

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root):
       return self.Height(root >=0)
    def Height(self, root):
        if not root: return 0
        left = self.Height(root.left)
        right = self.Height(root.right)
        if left < 0 or right < 0 or abs(right-left)>1: return -1
        return max(left, right) + 1


        

File: JavaScript/Easy/trees/balanced_binary_tree.js
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

const TreeNode = (val, left, right) => {
    this.val = (val === undefined ? 0: val)
    this.left = (left === undefined ? null: left)
    this.right = (right === undefined ? null: right)
}


var isBalanced = function (root) {
    if (root == null) {return true}
    if (Height(root) == -1) {return false;}
    return true
}

var Height = function(root) {
    if (root == null) {return 0}
    let left = Height(root.left)
    let right = Height(root.right)
    if (left < 0 || right < 0 || Math.abs(left-right)>1) {
        return -1
    }
    return Math.max(left, right) + 1
}